resolwe/flow/tests/test_models.py

```diff
@@ -521,6 +524,244 @@ def test_raw_process(self):
         self.assertNotEqual(response.data['id'], self.data.pk)
 
 
+class DuplicateTestCase(APITestCase):
+
+    def setUp(self):
+        super().setUp()
+
+        user_model = get_user_model()
+        self.user = user_model.objects.create(username='test_user', password='test_pwd')
+        self.duplication_user = user_model.objects.create(username='duplication_user', password='duplication_pwd')
+
+    def test_data_duplicate(self):
```

Comment:
There's a lot of properties to test when duplicating. Splitting the test into smaller parts would mean duplicating the objects in `setUp` and then testing different properties in each test. Is that what you meant?

Reply:
Yes, sometimes it can be quite hard to read a test if it is more than 100 lines long.

Other thing is that you should only test reasonable attributes, e.g.
1. It doesn't make sense to test the name of a copy of a copy as it is only copied and truncated
2. It does make sense to test relations of copy of a copy or even relations of a 2nd copy of the original object

Reply:
I don't know, tried breaking it up, but it's even messier. Creating data object in `setUp` seems confusing, because each test case needs to create its own specific objects. Sure, it's a long test, but at least everything's contained and separated by comments.

Creating another `TestCase` just for this and then having another `TestCase` to test other things (`test_input_rewiring`, `test_data_status_not_done`,..) seems like a unnecessary overhead.

Reply:
I don't see any benefit of checking every attribute of `duplicate_of_duplicate` in this test. In this case, I would prefer to split this test in two:

- create a minimal example of duplication and assert all attributes
- focus only on dependencies in the second and check what happens if you copy already copied object, what happens if you copy a source object twice,...
- you can add other tests for more complex parts of the process if there are any

I agree it is not the best thing to add a separate test case for each test, but maybe it would be beneficial to add a separate test case for all duplication tests.